### PR TITLE
Test Generation

### PR DESCRIPTION
**Pull request by Dargones**
# Overview

This pull request adds functionality that allows generating tests that (when possible) guarantee path- or block-coverage. Oracle is not provided. Detailed documentation can be found in this [README.md](https://github.com/Dargones/dafny/tree/PatchesToCounterExamples/Source/DafnyTestGeneration#readme) file included in the PR. Test generation supports all basic types, user-defined classes, as well as sequences, sets, and maps but does not yet support algebraic datatypes, arrays, and multisets. Below is a simple example of what test generation will do to a file called `Char.dfy` with the following code:
```
module Char {
  method compareToB(c: char) returns (i: int) {
    if (c == 'B') {
        return 0;
    } else if (c > 'B') {
        return 1;
    } else {
        return -1;
    }
  }
}
```
Running test generation like this:
```dafny /definiteAssignment:3 /testMode:Block Char.dfy ```
Will give the following list of tests (tabulation added manually):
```
include "Char.dfy"
module CharUnitTests {
  import Char
  method test0() returns (r0:int)  {
    r0 := Char.compareToB('!');
  }
  method test1() returns (r0:int)  {
    r0 := Char.compareToB(']');
  }
  method test2() returns (r0:int)  {
    r0 := Char.compareToB('B');
  }
}
```

# Changes
Aside from adding the `DafnyTestGeneration` and `DafnyTestGeneration.Test` projects to the solution, I made the following changes:
- Updated the workflow for XUnit tests to include the tests in `DafnyTestGeneration.Test` project.
- Updated `DafnyOptions.cs` and `DafnyDriver.cs` to parse and react to the new command line arguments described in the [README.md](https://github.com/Dargones/dafny/tree/PatchesToCounterExamples/Source/DafnyTestGeneration#readme) (`/testMode:<None|Block|Path|DeadCode>`, `/methodToTest:<method>`, `/seqLengthLimit:<n>`, and `/testInlineDepth:<n>`).
- Defined class `DafnyModelType`, instances of which are now used instead of strings to store the types of Dafny variables by the counterexample generation code in `DafnyServer`. This allows to effortlessly switch between a Dafny type name and its respective translation into Boogie (e.g. `Mo_dule_.Module2_.Cla__ss` and `Mo__dule___mModule2__.Cla____ss`).
- Added a few smaller fixes to counterexample generation code necessary to integrate it with test generation.
- Wrote a separate [README.md](https://github.com/Dargones/dafny/blob/PatchesToCounterExamples/Source/DafnyServer/CounterExampleGeneration/README.md) for counterexample generation.

### CODE REVIEW

File: Source/Dafny/DafnyOptions.cs [hunk 1]
```diff
@@ -101,6 +101,11 @@ public enum IncludesModes { None, Immediate, Transitive }
     public int Allocated = 3;
     public bool UseStdin = false;
     public bool ShowSnippets = false;
+    public enum TestModes { None, Block, Path };
```
Comment (keyboardDrummer) [inline Source/Dafny/DafnyOptions.cs:104]: Can you consider moving the code that was added to `DafnyOptions` to a class like `TestGenerationOptions` that includes the fields, the parsing logic and documentation, and then reference that from `DafnyOptions` ?

File: Source/DafnyLanguageServer.Test/Various/CounterExampleTest.cs [hunk 2]
```diff
@@ -505,9 +505,9 @@ method listHasSingleElement(list:List<real>)
       Assert.AreEqual(1, counterExamples.Length);
       Assert.AreEqual(1, counterExamples[0].Variables.Count);
       Assert.IsTrue(counterExamples[0].Variables.ContainsKey("s:seq<bv5>"));
-      Assert.AreEqual("(Length := 2, [1] := 2bv5)", counterExamples[0].Variables["s:seq<bv5>"]);
+      Assert.AreEqual("(Length := 2, [1] := 2)", counterExamples[0].Variables["s:seq<bv5>"]);
```
Comment (keyboardDrummer) [inline Source/DafnyLanguageServer.Test/Various/CounterExampleTest.cs:508]: Are there improvements to counterexample generation in this PR that could be factored out in a PR separate from the test generation?

File: Source/DafnyTestGeneration/README.md [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+[**Purpose**](#purpose) <br>
+[**General Approach**](#general-approach) <br>
+[**Implementation**](#implementation) <br>
+[**How to Generate Tests?**](#how-to-generate-tests) <br>
+[**How to Run Tests?**](#how-to-run-tests)
+
+## Purpose
+
+Block- and path-coverage tests of Dafny code can be used to:
```
Comment (keyboardDrummer) [inline Source/DafnyTestGeneration/README.md:11]: Could you elaborate more on these use-case? For example, how would I identify dead-code with this feature?

**Comment by Dargones** [comment]
Made 3 very small changes based on user-experience:

1) Tests are now printed line by line (so you don't have to wait until all tests are generated to see the output)
2) You can now pass the `/timeLimit` command line option
3) Type renaming of sequences, sets, and maps is now supported (e.g. `string32`)

**Review by RustanLeino — CHANGES_REQUESTED** [review]
Looks good. I only have a few minor remarks.

File: Source/DafnyTestGeneration.Test/Varia.cs [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+class Paths {
+  static method eightPaths (i:int)
+    returns (divBy2:bool, divBy3:bool, divBy5:bool)
+  {
+		if (i % 2 == 0) {
+			divBy2 := true;
+		} else {
+			divBy2 := false;
+		}
+		if (i % 3 == 0) {
```
[the remark this answers is not in the document]
Comment (Dargones) [reply Source/DafnyTestGeneration.Test/Varia.cs]: Thank you for catching that! Have replaced those with spaces.

File: Source/DafnyTestGeneration/README.md [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+        e.printStackTrace();
+    }
+    return parameters;
+}
+```
+
+## Example
+
+Suppose you have a file called `Char.dfy` with the following code:
+```
```
[the remark this answers is not in the document]
Comment (Dargones) [reply Source/DafnyTestGeneration/README.md]: Done! Didn't know one can do that with Dafny too.

File: Source/DafnyTestGeneration/README.md [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+- [BlockBasedModifier.cs](BlockBasedModifier.cs) and [PathBasedModifier.cs](PathBasedModifier.cs) - subclasses of `ProgramModifier` that generate instances of `ProgramModification` by inserting assertions that fail when a particular block is visited or a particular path is taken, respectively.
+- [DafnyInfo.cs](DafnyInfo.cs) - takes a Dafny Program and extracts certain information about it that is then used in `TestMethod.cs` and `Main.cs`
+
+## How to Generate Tests?
+
+- Test generation currently works with all basic types, user-defined classes, sequences, sets, and maps. It does not work with datatypes, arrays, and multisets. It is also not possible to generate tests for constructors.
+- To generate block- or path-coverage tests use the `/testMode:Block` or `/testMode:Path` arguments respectively. You will likely also need `definiteAssignment:3`, which will endure that you generate tests even for those methods that do not contain any assertions or pre-\post-conditions.
```
[the remark this answers is not in the document]
Comment (Dargones) [reply Source/DafnyTestGeneration/README.md]: Thank you for noticing the typo - fixed that in the latest batch of commits

**Comment by Dargones** [comment]
Thank you, Rustan! Have adopted all the changes you suggested.

File: Source/DafnyTestGeneration/README.md [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+- If you are using `/testTargetMethod` and would like to inline methods that are called from the method of interest, you can do so by setting `/testInlineDepth` to something larger than zero (zero is the default). The `/verifyAllModules` argument might also be relevant if the methods to be inlined are defined in included files.
+- To deal with loops, you should use `/loopUnroll` and also `/testSeqLengthLimit`. The latter argument adds an axiom that limits the length of any sequence to be no greater than some specified value. This restriction can be used to ensure that the number of loop unrolls is sufficient with respect to the length of any input sequence but it can also cause the program to miss certain corner cases.
+- The`/testMode:DeadCode` argument will make Dafny identify potential dead code in the specified file. Note that false negatives are possible if `/loopUnroll` is not used. False positives are also possible for a variety of reasons, such as `/loopUnroll` being assigned not high enough value.
```
Comment (robin-aws) [inline Source/DafnyTestGeneration/README.md:46]: I get that generating tests and identifying dead code share a common implementation, but I think they should be exposed as a very different modes of operation. The former produces test files, whereas the latter produces warnings (and I'm hoping we can make it a feature of the language server instead soon).

I would suggest a separate `/warnDeadCode` option, similar to the existing `/warnShadowing`

File: Source/DafnyTestGeneration/README.md [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+- [BlockBasedModifier.cs](BlockBasedModifier.cs) and [PathBasedModifier.cs](PathBasedModifier.cs) - subclasses of `ProgramModifier` that generate instances of `ProgramModification` by inserting assertions that fail when a particular block is visited or a particular path is taken, respectively.
+- [DafnyInfo.cs](DafnyInfo.cs) - takes a Dafny Program and extracts certain information about it that is then used in `TestMethod.cs` and `Main.cs`
+
+## How to Generate Tests?
+
+- Test generation currently works with all basic types, user-defined classes, sequences, sets, and maps. It does not work with datatypes, arrays, and multisets. It is also not possible to generate tests for constructors.
+- To generate block- or path-coverage tests use the `/testMode:Block` or `/testMode:Path` arguments respectively. You will likely also need `definiteAssignment:3`, which will ensure that you generate tests even for those methods that do not contain any assertions or pre-\post-conditions.
```
Comment (robin-aws) [inline Source/DafnyTestGeneration/README.md:42]: Let's use `generateTest...` instead of just `/test...` as the common prefix for these options, to avoid implying that we're actually testing or running tests. If and when we separate out the different `dafny` commands as suggested in https://github.com/dafny-lang/dafny/issues/1322#issuecomment-889023432, we can make this feature a separate `dafny generate-tests ...` command.

File: Source/DafnyTestGeneration/README.md [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+- To deal with loops, you should use `/loopUnroll` and also `/testSeqLengthLimit`. The latter argument adds an axiom that limits the length of any sequence to be no greater than some specified value. This restriction can be used to ensure that the number of loop unrolls is sufficient with respect to the length of any input sequence but it can also cause the program to miss certain corner cases.
+- The`/testMode:DeadCode` argument will make Dafny identify potential dead code in the specified file. Note that false negatives are possible if `/loopUnroll` is not used. False positives are also possible for a variety of reasons, such as `/loopUnroll` being assigned not high enough value.
+
+## How to Run Tests?
+
+To run the tests, you would first need to compile the code to some other language. Once this is done, the main difficulty is to mock the arguments that the test methods take. The following code (to be inserted into the class with the tests) should do the job in Java:
```
Comment (robin-aws) [inline Source/DafnyTestGeneration/README.md:50]: Is it possible to add another example that ends up producing test methods that take arguments? It's a bit hard to follow this since the example immediately below doesn't.

File: Source/DafnyTestGeneration/README.md (hunk 8 above)
Comment (robin-aws) [inline Source/DafnyTestGeneration/README.md:42]: Why is `definiteAssignment:3` necessary exactly? Does it prevent such methods from being completely erased in the translation to Boogie?

**Review by robin-aws — CHANGES_REQUESTED** [review]
Looks great! I'm happy to defer to Rustan and Remy on the deeper review, my only real concern is with the API for using the new feature.

File: Source/DafnyTestGeneration/README.md [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+Running test generation like this:
+
+```dafny /definiteAssignment:3 /testMode:Block Char.dfy ```
+
+Will give the following list of tests (tabulation added manually):
+```
+include "Char.dfy"
+module CharUnitTests {
+  import Char
+  method test0() returns (r0:int)  {
```
[the remark this answers is not in the document]
Comment (Dargones) [reply Source/DafnyTestGeneration/README.md:103]: Would `{:test}` compile successfully if used on methods that take parameters? (this is necessary to mock objects as outlined in "How to run tests" section in the [README](https://github.com/Dargones/dafny/tree/PatchesToCounterExamples/Source/DafnyTestGeneration#readme)). I agree that it would be great to use the annotation but I am wondering if it would be better to create a separate PR about extending the way the compilers treat `{:test}` before using it in this way.

[the remark this answers is not in the document]
Comment (robin-aws) [reply Source/DafnyTestGeneration/README.md:103]: No, it wouldn't work if the methods take parameters without more work, even in the C# case. `[Fact]` only works if there are no parameters. `[Theory]` can be used to support parameterized tests, but you'd have to supply a separate `DataAttribute` of some kind to provide the mocked object arguments.

No worries if you want to ignore this suggestion for now - it may be better to follow up on it once Dafny's support for tests is fleshed out more.

File: Source/DafnyTestGeneration/README.md (hunk 7 above)
Comment (Dargones) [reply Source/DafnyTestGeneration/README.md]: Done! It would indeed be great if one could in the future highlight potentially dead code directly in the IDE using the LanguageServer.

File: Source/DafnyTestGeneration/README.md (hunk 8 above)
Comment (Dargones) [reply Source/DafnyTestGeneration/README.md]: Have made the change in one of my latest commits. Separating this into a standalone `generate-tests` command would be great for reducing argument name length.

File: Source/DafnyTestGeneration/README.md (hunk 9 above)
Comment (Dargones) [reply Source/DafnyTestGeneration/README.md]: Changed the example in one of my latest commits to show how to work around test methods that take arguments.

File: Source/DafnyTestGeneration/README.md (hunk 8 above)
Comment (Dargones) [reply Source/DafnyTestGeneration/README.md]: Yes, Dafny to Boogie translator would sometimes leave out the implementation and only emit the procedure - using `definiteAssignment:3` typically prevents this from happening. I have updated the explanation in the README in one of my latest commits.

**Comment by Dargones** [comment]
Thank you for all the feedback, @robin-aws ! I am a bit hesitant about using the `{:test}` annotation given that these methods might take arguments but I am happy to make the change if you think it would be OK. Aside from this, I have adopted all your suggestions. Let me know if there is anything else I should do!

File: Source/DafnyTestGeneration.Test/Varia.cs [hunk 11]
```diff
@@ -0,0 +1,329 @@
+using System.Linq;
+using System.Text.RegularExpressions;
+using Microsoft.Dafny;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+namespace DafnyTestGeneration.Test {
+
+
+  [TestClass]
+  public class Varia {
```
Comment (robin-aws) [inline Source/DafnyTestGeneration.Test/Varia.cs:10]: Typo?

**Review by robin-aws — CHANGES_REQUESTED** [review]
Thanks for the updates @Dargones! Don't worry about `{:test}` for now, we'll come back to that once Dafny tests are more deeply supported and we have a solution for test method parameters.

My only remaining concern is end-to-end testing: I really appreciate the unit test coverage you've provided with the `DafnyTestGeneration.Test` package, but I think we need at least one lit test that generates and actually executes a test file, as you describe in the README.md. https://github.com/dafny-lang/dafny/blob/master/Test/DafnyTests/TestAttribute.dfy is somewhat similar and might give you an idea of how to do such a multi-command test.

File: Source/DafnyTestGeneration.Test/Varia.cs (hunk 11 above)
Comment (Dargones) [reply Source/DafnyTestGeneration.Test/Varia.cs]: Changed to "Various" - thank you!

File: Test/DafnyTestGeneration/TestGeneration.dfy.expect [hunk 12]
```diff
@@ -0,0 +1,16 @@
+include "TestGeneration.dfy"
+module TestGenerationUnitTests {
+import M
+method test0(v0:M.Value) returns (r0:int)  modifies v0 {
+v0.v := -INT;
```
Comment (robin-aws) [inline Test/DafnyTestGeneration/TestGeneration.dfy.expect:5]: This doesn't look right - was `INT` supposed to be substituted?

Comment (Dargones) [reply Test/DafnyTestGeneration/TestGeneration.dfy.expect:5]: It is, since there is no way of telling which particular integer the solver will come up with. I am actually modifying the test right now to execute this dafny code - will let you know once I succeed.

**Comment by Dargones** [comment]
Thank you @robin-aws ! I have now added a lit test that executes the generated Dafny tests in Java and compares the results. For some reason, GitHub does not seem to run this test though (it only runs 223 of 1115 lit tests for that matter). Is there some setting I need to add that would allow GitHub to execute this test?

**Comment by sorawee** [comment]
IIUC, tests are divided into multiple runs in order to parallelize. Your test would be in one of the `(ubuntu-latest, 1)`, `(ubuntu-latest, 2)`, ..., `(ubuntu-latest, 5)`, I think.

**Comment by Dargones** [comment]
Thank you, @sorawee! You are right, it is in `build (ubuntu-latest, 1)` and passes, so all is well!

**Review by robin-aws — DISMISSED** [review]
Thanks for adding the lit test, and congrats on by far the most complicated lit program I've seen yet! :)

File: Source/DafnyTestGeneration/README.md [hunk 13]
```diff
@@ -77,47 +77,58 @@ public static Object[] mockParameters(Method method) {
...
+module M {
+    class Value {
+        var v:int;
+    }
+    method compareToZero(v:Value) returns (i:int) {
+        if (v.v == 0) {
+            return 0;
+        } else if (v.v > 1) {
+            return 1;
+        }
```
Comment (RustanLeino) [inline Source/DafnyTestGeneration/README.md]: Use 2 spaces for indent. Same comment for L102-L116 below.

File: Test/DafnyTestGeneration/TestGeneration.dfy [hunk 14]
```diff
@@ -0,0 +1,17 @@
...
+        var v:int;
+    }
+    method compareToZero(v:Value) returns (i:int) {
+        if (v.v == 0) {
+            return 0;
+        } else if (v.v > 0) {
+            return 1;
+        }
+        return -1;
+    }
```
Comment (RustanLeino) [inline Test/DafnyTestGeneration/TestGeneration.dfy:30]: Use 2 spaces for indentation.

File: Test/DafnyTestGeneration/TestGeneration.dfy [hunk 15]
```diff
@@ -1,5 +1,19 @@
...
+// RUN: cp TestGeneration.dfy %t.dfy
+// RUN: %dafny /definiteAssignment:3 /generateTestMode:Block %t.dfy > %t-tests.dfy
+
+// Compiling test to java:
+// RUN: cd Output; %dafny /compileTarget:java %t-tests.dfy
+
+// Adding reflection code that allows running the tests:
+// RUN: cp ../import.txt .
+// RUN: cp ../reflectionCode.txt .
+// RUN: perl -pe 's/import M_Compile.*;/`cat import.txt`/ge' -i %t-tests-java/TestGenerationUnitTests_Compile/__default.java
```
Comment (RustanLeino) [inline Test/DafnyTestGeneration/TestGeneration.dfy:11]: Are `cp` and `perl` available on Windows?

File: Source/DafnyTestGeneration/README.md [hunk 16]
```diff
@@ -39,7 +39,7 @@ Even though it is possible to run the tests in this way, we currently provide no
 ## How to Generate Tests?
 
 - Test generation currently works with all basic types, user-defined classes, sequences, sets, and maps. It does not work with datatypes, arrays, and multisets. It is also not possible to generate tests for constructors.
-- To generate block- or path-coverage tests use the `/generateTestMode:Block` or `/generateTestMode:Path` arguments respectively. You will likely also need `definiteAssignment:3`, which will ensure that you generate tests even for those methods that do not contain any assertions or pre-\post-conditions.
+- To generate block- or path-coverage tests use the `/generateTestMode:Block` or `/generateTestMode:Path` arguments respectively. You will likely also need `definiteAssignment:3`, which forces the Dafny-to-Boogie translator to emit implementations even for those methods that don't have any assertions or pre-\post-conditions. This, in turn, allows to generate tests for such methods.
```
Comment (RustanLeino) [inline Source/DafnyTestGeneration/README.md:42]: This is not true. Even with `/definiteAssignment:3`, Dafny still performs the optimization of not generating a Boogie procedure if there are no proof obligations. The reason you may think the two have anything to do with each other is that `/definiteAssignment:3` causes there to be more proof obligations. Still, a method like

``` dafny
method M() {
}
```

does not generate any Boogie procedure.

I suggest you remove the mention of `/definiteAssignment:3`. Instead, you could just remark that test generation relies on Dafny to generate Boogie procedures and that Dafny does not generate a Boogie procedure when there are no proof obligations, and thus no tests will be generated for those. Alternatively, you could disable this optimization when the test-generation flags are used (see the `EmitImplementation` method in `Translator.cs`).

**Review by RustanLeino — CHANGES_REQUESTED** [review]

File: Source/DafnyTestGeneration/README.md (hunk 13 above)
Comment (Dargones) [reply Source/DafnyTestGeneration/README.md]: Done!

File: Test/DafnyTestGeneration/TestGeneration.dfy (hunk 14 above)
Comment (Dargones) [reply Test/DafnyTestGeneration/TestGeneration.dfy]: Thank you for catching this, have made the change

File: Test/DafnyTestGeneration/TestGeneration.dfy (hunk 15 above)
Comment (Dargones) [reply Test/DafnyTestGeneration/TestGeneration.dfy]: Not by default, unfortunately. Added a `Test/lit.site.cfg` that should disable this test for Windows

File: Source/DafnyTestGeneration/README.md (hunk 16 above)
Comment (Dargones) [reply Source/DafnyTestGeneration/README.md]: Thank you for clarifying this! I have paraphrased the explanation as you suggested. 

My reasoning was that one would not generate tests for a method that doesn't take any arguments and that `definiteAssignemnt:3` adds proof obligations for all methods that do (although I might be wrong about this) but I do realize that it technically has its own purpose and that one does not necessarily want to use it. In any case, I no longer mention `definiteAssignemnt:3` in the description (although I still have to use it in the example and the lit test)

**Review by RustanLeino — APPROVED** [review]
